orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,272 @@
+	emptyEC := EndpointCriteria{}
+	if bf.currentEndpoint.Endpoint == emptyEC.Endpoint {
+		// first time allocation
+		bf.Logger.Info("Set First Endpoint")
+		bf.currentEndpoint = bf.Endpoints[0]
+		bf.LastShuffledAt = time.Now()
+	}
+	// otherwise check if it is time to shuffle or not
+	if bf.isShuffleTime() {
+		bf.shuffleEndpoint()
+	}
+
+}
+
+func (bf *BlockFetcher) getIdxForEndpoint(e string) (int, error) {
+	var err error
+
+	for idx, e := range bf.Endpoints {
+		if bf.currentEndpoint.Endpoint == e.Endpoint {
+			return idx, err
+		}
+	}
+	return 0, errors.New("Endpoint not in the list")
+}
+
+func (bf *BlockFetcher) shuffleEndpoint() error {
+	// Iterate through the bf.Endpoints and choose the next endpoint
+	// which is not suspected
+	// shuffle endpoint if there is one
+
+	currIdx, err := bf.getIdxForEndpoint(bf.currentEndpoint.Endpoint)
+	if err != nil {
+		return nil
+	}
+	// Lets us increase the current idx by 1, for shuffle
+	// shuffle here is just iterating to next endpoint
+	// increment is circular, i.e after last element, an increment will lead to the first element.
+
+	nextIdx := (currIdx + 1) % len(bf.Endpoints)
+	// check nextIdx in suspect list
+
+	for {
+
+		// TODO: keep a bound on this loop so that it doesn't go infinite
+		if !bf.suspects.has(bf.Endpoints[nextIdx].Endpoint) {
+			bf.currentEndpoint = bf.Endpoints[nextIdx]
+			break
+		}
+		nextIdx = (currIdx + 1) % len(bf.Endpoints)
+	}
+
+	bf.LastShuffledAt = time.Now()
+	return nil
+}
+
+func (bf *BlockFetcher) isShuffleTime() bool {
+	bf.Logger.Debug("Check Shuffle Time")
+	if bf.ShuffleTimeout == time.Duration(0) {
+		// It should return only false if Time based Shuffling is not enabled
+		bf.Logger.Debugf("Shuffle Disabled.")
+		return false
+	}
+
+	shuffle := time.Now().After(bf.LastShuffledAt.Add(bf.ShuffleTimeout))
+	bf.Logger.Debugf("Shuffle Enabled. Time to shuffle %v", shuffle)
+	return shuffle
+}
+
+func (bf *BlockFetcher) isEndpointByzantine(ec EndpointCriteria) bool {
+	// probe the other endpoints
+
+	// choose (MaxByzantineFaults + 1) endpoints other than currentEndpoint
+	//currIdx, err := bf.getIdxForEndpoint(bf.currentEndpoint.Endpoint)
+	currIdx, _ := bf.getIdxForEndpoint(ec.Endpoint)
+	// TODO: handle above error case
+
+	endpointsToProbe := []EndpointCriteria{}
+	for i := int64(currIdx + 1); i <= (bf.MaxByzantineFaults + int64(currIdx)); i++ {
+		// add the next (bf.MaxByzantineFaults + 1)
+		endpointsToProbe = append(endpointsToProbe, bf.Endpoints[i%(int64(len(bf.Endpoints)))])
+	}
+
+	// add a Sync.WaitGroup and probe using go routines
+	waitGrp := sync.WaitGroup{}
+	// recieve the blocks using channels to a slice
+
+	// verify the slice and decide whether current node had byzantine behavior
+	type bps struct {
+		block *common.Block
+		ec    EndpointCriteria
+	}
+
+	bpsInfo := make(chan *bps, len(endpointsToProbe))
+
+	// start probing
+	for _, ec := range endpointsToProbe {
+		// get a puller
+		waitGrp.Add(1)
```

Comment:
it's better to `waitGrp.Add(len(endpointsToprobe)` before the for loop.

Reply:
agree